server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -303,6 +349,67 @@ static boolean execute(SearchContext searchContext,
         }
     }
 
+     private static Query tryRewriteNumericLongOrDateSort(SearchContext searchContext, IndexReader reader) throws IOException {
+        // child docs can inherit scores from parent, so avoid has_parent query
+        if (searchContext.request().source().query().getName().equals("has_parent")) return null;
+
+        if (searchContext.searchAfter() != null) return null;
+        if (searchContext.scrollContext() != null) return null;
+        if (searchContext.collapse() != null) return null;
+        Sort sort = searchContext.sort().sort;
+        if (Sort.RELEVANCE.equals(sort)) return null;
+        if (Sort.INDEXORDER.equals(sort)) return null;
+        SortField sortField = sort.getSort()[0];
+
+         // check if this is a field of type Long or Date, that is indexed and has doc values
+        String fieldName = sortField.getField();
+        if (fieldName == null) return null; // happens when _score is the 1st sort field
+        if (searchContext.mapperService() == null) return null; // mapperService can be null in tests
+        final MappedFieldType fieldType = searchContext.mapperService().fullName(fieldName);
+        if (fieldType == null) return null; // for unmapped fields, default behaviour depending on "unmapped_type" flag
+         if ((fieldType.typeName().equals("long") == false) && (fieldType instanceof DateFieldType == false)) return null;
+        if (fieldType.indexOptions() == IndexOptions.NONE) return null; //TODO: change to pointDataDimensionCount() when implemented
+        if (fieldType.hasDocValues() == false) return null;
+
+        // check that setting of missing values allows optimization
+        if (sortField.getMissingValue() == null) return null;
+        Long missingValue = (Long) sortField.getMissingValue();
+        boolean missingValuesAccordingToSort = (sortField.getReverse() && (missingValue == Long.MIN_VALUE)) ||
+        ((sortField.getReverse() == false) && (missingValue == Long.MAX_VALUE));
+        if (missingValuesAccordingToSort == false) return null;
+
+         // check for multiple values
+        if (PointValues.size(reader, fieldName) != PointValues.getDocCount(reader, fieldName)) return null; //TODO: handle multiple values
+
+        byte[] minValueBytes = PointValues.getMinPackedValue(reader, fieldName);
+        byte[] maxValueBytes = PointValues.getMaxPackedValue(reader, fieldName);
+        if ((maxValueBytes == null) || (minValueBytes == null)) return new MatchAllDocsQuery(); // no values on this shard
```

Comment:
Apologies for the back-and-forth, I just realized that returning a query that produces constant-scores would _not_ help because when there are multiple sorting criteria, we can only ask hits to have a score that is greater than *or equal to* to best k-th score, rather than strictly greater to like when sorting only by score.